coast/gridded.py

```diff
@@ -0,0 +1,865 @@
+            e3u_new[:, :, :, :-1] = e3u_temp
+            e3u_new[:, :, :, -1] = ds_dom.e3u_0[:, :, -1]
+            e3u_new['longitude'] = ds_dom.glamu
+            e3u_new['latitude'] = ds_dom.gphiu
+            if e3u:
+                e3_return.append(e3u_new.squeeze())
+
+        # area averaged interpolation onto the u-grid to get e3v 
+        if e3v:
+            e1e2v = ds_dom.e1v * ds_dom.e2v
+            e3v_temp = (((0.5 / e1e2v[:-1, :]) * ((e1e2t[:-1, :] * e3t_dt[:, :, :-1, :])
+                                                  + (e1e2t[1:, :] * e3t_dt[:, :, 1:, :])))
+                        .transpose('t_dim', 'z_dim', 'y_dim', 'x_dim'))
+            e3v_temp = e3v_temp.where(e3t_dt[:, :, 1:, :] != 0, 0)
+            e3v_temp = e3v_temp.where(e3v_temp.z_dim < ds_dom.bottom_level[:-1, :], 0)
+            e3v_temp = (e3v_temp + ds_dom.e3v_0[:, :-1, :])
+            e3v_new = xr.zeros_like(e3t_new)
+            e3v_new[:, :, :-1, :] = e3v_temp
+            e3v_new[:, :, -1, :] = ds_dom.e3v_0[:, -1, :]
+            e3v_new['longitude'] = ds_dom.glamv
+            e3v_new['latitude'] = ds_dom.gphiv
+            e3_return.append(e3v_new.squeeze())
+
+        # area averaged interpolation onto the u-grid to get e3f
+        if e3f:
+            e1e2f = ds_dom.e1f * ds_dom.e2f
+            e3u_dt = e3u_new - ds_dom.e3u_0
+            e3f_temp = (((0.5 / e1e2f[:-1, :]) * ((e1e2u[:-1, :] * e3u_dt[:, :, :-1, :])
+                                                  + (e1e2u[1:, :] * e3u_dt[:, :, 1:, :])))
+                        .transpose('t_dim', 'z_dim', 'y_dim', 'x_dim'))
+            e3f_temp = e3f_temp.where(e3u_dt[:, :, 1:, :] != 0, 0)
+            e3f_temp = e3f_temp.where(e3f_temp.z_dim < ds_dom.bottom_level[:-1, :], 0)
+            e3f_temp = (e3f_temp + ds_dom.e3f_0[:, :-1, :])
+            e3f_new = xr.zeros_like(e3t_new)
+            e3f_new[:, :, :-1, :] = e3f_temp
+            e3f_new[:, :, -1, :] = ds_dom.e3f_0[:, -1, :]
+            e3f_new['longitude'] = ds_dom.glamf
+            e3f_new['latitude'] = ds_dom.gphif
+            e3_return.append(e3f_new.squeeze())
+
+        # simple vertical interpolation for e3w. Special treatment of top and bottom levels   
+        if e3w:
+            # top levels correction same at e3t
+            e3w_new = (ds_dom.e3w_0 + e3t_dt).transpose('t_dim', 'z_dim', 'y_dim', 'x_dim')
+            # levels between top and bottom
+            e3w_new[dict(z_dim=slice(1, None))] = (0.5 * e3t_dt[:, :-1, :, :]
+                                                   + 0.5 * e3t_dt[:, 1:, :, :]
+                                                   + ds_dom.e3w_0[1:, :, :])
+            # bottom and below levels
+            e3w_new = e3w_new.where(e3w_new.z_dim < ds_dom.bottom_level, e3t_dt.shift(z_dim=1) + ds_dom.e3w_0)
+            e3_return.append(e3w_new.squeeze())
+
+        return tuple(e3_return)
+
+    def harmonics_combine(self, constituents, components=['x', 'y']):
```

Comment:
Ideally we should avoid having mutable objects as defaults. Could this be a tuple?